adapters/lora-vertx/src/main/java/org/eclipse/hono/adapter/lora/impl/LoraProtocolAdapter.java

```diff
@@ -243,6 +269,106 @@ void handleProviderRoute(final HttpContext ctx, final LoraProvider provider) {
+            .compose(success ->
+                getRegistrationClient().assertRegistration(command.getTenant(), command.getGatewayId(), null,
+                    commandContext.getTracingContext()))
+            .compose(registrationAssertion -> {
+                final CompletableFuture<Boolean> booleanCompletableFuture = onCommandReceived(command, subscriptionKey,
+                    registrationAssertion);
+                final CompletableFuture<Future<Void>> future =
+                    booleanCompletableFuture.handle((aBoolean, throwable) -> {
+                        if (throwable != null) {
+                            return Future.failedFuture(throwable);
+                        }
+                        if (aBoolean) {
+                            return Future.succeededFuture();
+                        }
+                        return Future.failedFuture("Failure while sending command");
+                    });
+                booleanCompletableFuture.join();
+                try {
+                    return future.get();
+                } catch (Exception e) {
+                    return Future.failedFuture(e);
+                }
+            })
+            .onComplete(aVoid -> {
+                addMicrometerSample(commandContext, timer);
+                commandContext.accept();
+                metrics.reportCommand(
+                    command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                    subscriptionKey.getTenant(),
+                    tenantTracker.result(),
+                    MetricsTags.ProcessingOutcome.FORWARDED,
+                    command.getPayloadSize(),
+                    timer);
+            })
+            .onFailure(t -> {
+                commandContext.release();
+                metrics.reportCommand(
+                    command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                    subscriptionKey.getTenant(),
+                    tenantTracker.result(),
+                    MetricsTags.ProcessingOutcome.from(t),
+                    command.getPayloadSize(),
+                    timer);
+            });
+    }
+
+    private CompletableFuture<Boolean> onCommandReceived(final Command command,
+                                                         final SubscriptionKey subscriptionKey, final RegistrationAssertion registrationAssertion) {
+        final CompletableFuture<Boolean> completableFuture = new CompletableFuture<>();
+        final LoraProvider loraProvider = commandSubscriptions.get(subscriptionKey);
+        final CommandEndpoint commandEndpoint = registrationAssertion.getCommandEndpoint();
+        final Optional<LoraCommand> optionalLoraCommand = loraProvider.getCommand(commandEndpoint, command.getDeviceId(),
+            command.getPayload());
+        if (optionalLoraCommand.isPresent()) {
+            final LoraCommand loraCommand = optionalLoraCommand.get();
+            final HttpClient httpClient = vertx.createHttpClient();
```

Comment:
This client instance should better be reused for multiple requests in this adapter instance.